tests/unit/bootstrap_record_test.cc

```diff
@@ -0,0 +1,190 @@
+#include <seastar/fs/block_device.hh>
+#include <seastar/fs/temporary_file.hh>
+#include <seastar/core/thread.hh>
+#include <seastar/core/units.hh>
+#include <seastar/testing/test_case.hh>
+#include <seastar/testing/thread_test_case.hh>
+#include <seastar/testing/test_runner.hh>
+
+using namespace seastar;
+using namespace seastar::fs;
+
+namespace {
+
+constexpr off_t device_size = 1 * MB;
+
+class mock_block_device_impl : public block_device_impl {
+public:
+    using buf_type = basic_sstring<uint8_t, size_t, 32, false>;
+    buf_type buf;
+    ~mock_block_device_impl() override = default;
+
+    future<size_t> write(uint64_t pos, const void* buffer, size_t len, const io_priority_class& pc) override {
+        if (buf.size() < pos + len)
+            buf.resize(pos + len);
+        std::memcpy(buf.data() + pos, buffer, len);
+        return make_ready_future<size_t>(len);
+    }
+
+    future<size_t> read(uint64_t pos, void* buffer, size_t len, const io_priority_class& pc) override {
+        std::memcpy(buffer, buf.data() + pos, len);
+        return make_ready_future<size_t>(len);
+    }
+
+    future<> flush() override {
+        return make_ready_future<>();
+    }
+
+    future<> close() override {
+        return make_ready_future<>();
+    }
+};
+
+future<block_device> prepare_device() {
+    return async([&] {
+        sstring device_path = [&]() -> sstring {
+            temporary_file tmp_device_file("/tmp/bootstrap_record_test_file");
+            return tmp_device_file.path();
+        }();
+
+        file f = open_file_dma(device_path, open_flags::rw | open_flags::create).get0();
+        f.truncate(device_size).get();
+        f.close().get();
+
+        return open_block_device(device_path).get0();
+    });
+}
+
+}
+
+SEASTAR_TEST_CASE(test_valid_read_write) {
+    return prepare_device().then([] (block_device dev) {
+        return do_with(std::move(dev), [] (block_device& dev) {
+            return async([&] {
+                // basic test
+                bootstrap_record write_record(1, 2, 3, 4, 3, {{6, {7, 8}}, {9, {10, 11}}, {12, {13, 14}}});
+                write_record.write_to_disk(dev).get();
+                auto read_record = bootstrap_record::read_from_disk(dev).get0();
+                BOOST_REQUIRE_EQUAL(write_record.cluster_size, read_record.cluster_size);
+
+                // max number of shards
+                write_record.shards_nb = bootstrap_record::max_shards_nb;
+                write_record.shards_info.resize(write_record.shards_nb, {1, {2, 3}});
+                read_record = bootstrap_record::read_from_disk(dev).get0();
+                BOOST_REQUIRE_EQUAL(write_record.cluster_size, read_record.cluster_size);
+
+                // 0 shards
+                write_record.shards_nb = 0;
+                write_record.shards_info.clear();
+                read_record = bootstrap_record::read_from_disk(dev).get0();
+                BOOST_REQUIRE_EQUAL(write_record.cluster_size, read_record.cluster_size);
```

Comment:
Missing writes of the changed write_record to disk, you probably want to compare changed values instead of cluster_size too.